feat: ✨ Add Leave Lobby Functionality

mrjoshua520 opened the pull request:
closes #39

mrjoshua520 committed:
feat(functions): ✨ Create structure for the leave lobby function

This adds the basic structure and imports for the leave lobby function

mrjoshua520 committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into jsalkeld/leave-lobby

mrjoshua520 committed:
feat(functions): ✨ Initial leave lobby functionality

This commit removes the proper UID from the uids array in the lobby collection. Removing the player from the players array is not functional

mrjoshua520 committed:
feat(functions): ✨ Second half of leave lobby functions

User is now removed from the players array as well as the uid array

mrjoshua520 committed:
feat(client): ✨ Leave button redirect to splash page

The leave lobby button no redirects the user to the splash page after removing them from the lobby

mrjoshua520 committed:
fix(client): 🐛 Start game button no longer shows for non-host players

mrjoshua520 committed:
feat(functions): ✨ Delete the lobby if the last player leaves

mrjoshua520 committed:
feat(client): ✨ Alert user when they close the tab

Alerts user when they close the tab. If they are in the lobby remove them assuming they decided to leave

mrjoshua520 committed:
feat(client): ✨ Errors display to the user

Displays start game and leave lobby errors to the user

mrjoshua520 committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into jsalkeld/leave-lobby

mrjoshua520 committed:
fix(client): 🐛 Fix host not seeing start game button

mrjoshua520 committed:
fix(client): 🐛 Fix svelte check errors on imports

nstringham commented:
while playing the game if you reload the page then you leave the lobby. is that acceptable behavior?

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -35,6 +39,23 @@
+    try {
+      await findAndStartLobby(lobbyCode);
+    } catch (err) {
+      console.log("HELLLO" + err);
```
I doubt you intended to leave that in

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -44,8 +65,21 @@
-      <Button on:click={() => startGame({ code: lobbyCode })}><Label>Start Game</Label></Button>
+    {#if auth.currentUser?.uid === lobby.uids[0]}<div class="start">
+        <Button on:click={() => startGame()}><Label>Start Game</Label></Button>
+      </div>{/if}
```
I feel like it would look better if it was like
```svelte
{#if auth.currentUser?.uid === lobby.uids[0]}
  <div class="start">
    <Button on:click={() => startGame()}><Label>Start Game</Label></Button>
  </div>
{/if}
```

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -44,8 +65,21 @@
+      <Button
+        on:click={() => {
+          leaveLobby();
+          goto("/");
```
this makes it go to the home page before it is done leaving the lobby because leaveLobby is not awaited

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -44,8 +65,21 @@
+    </div>
+    <div class="error">
+      {#if errorMessage !== ""}
+        <p class="error">{errorMessage}</p>
```
you probobly don't want 2 things with the same class like that

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -60,7 +94,9 @@
+  .error {
     display: grid;
     place-items: center;
   }
```
if these 3 classes are the same then you should only have one class that all 3 elements share

nstringham commented on client/src/lib/firebase/start-lobby.ts:
```diff
@@ -0,0 +1,11 @@
+
+  // make request to server
+  await startGame({ code: id });
+}
```
don't really need this but whatever

nstringham commented on client/src/lib/firebase/leave-lobby.ts:
```diff
@@ -0,0 +1,11 @@
+
+  // make request to server
+  await leaveLobby({ code: id });
+}
```
don't really need this but whatever

nstringham commented on client/src/routes/game/+page.svelte:
```diff
@@ -64,6 +65,13 @@
+
+    addEventListener("beforeunload", (event) => {
+      event.returnValue = "\\o/";
+      if (lobby?.state === "WAIT" && lobbyCode !== null) {
```
this code always runs even if the user clicks "cancel" maybe use the "unload" event

mrjoshua520 replied:
We decided not to have the user leave during the unload event so this won't need to be addressed

mrjoshua520 committed:
refactor(client): 🔥 Remove unneeded files

Remove unneeded files and changes functions accordingly. Clean up some CSS. Remove an unneeded console log

mrjoshua520 committed:
refactor(client): 🐛 Fix leaving the lobby before the user actually leaves the lobby

mrjoshua520 committed:
refactor(client): ♻️ Change addeventlistener to svelte:window

mrjoshua520 committed:
refactor(client): 🔥 Removes unused import statements

nstringham commented on functions/src/lobby.ts:
```diff
@@ -140,6 +140,48 @@ export const joinLobby = functions.https.onCall((data: unknown, context): Promis
+
+    // If the last player is leaving delete the document instead
+    if (uids.length === 1) {
+      await transaction.delete(lobby);
```
this await doesn't do anything because `transaction.delete` does not return a promise

nstringham replied:
this await doesn't do anything because `transaction.delete` does not return a promise

nstringham commented on functions/src/lobby.ts:
```diff
@@ -140,6 +140,48 @@ export const joinLobby = functions.https.onCall((data: unknown, context): Promis
+      await transaction.delete(lobby);
+    } else {
+      // Remove player from the lobby
+      await transaction.update(lobby, {
```
this await doesn't do anything because `transaction.update` does not return a promise

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -44,8 +63,22 @@
     <SelectAvatar {lobby} {lobbyCode} />
-    <div class="start">
-      <Button on:click={() => startGame({ code: lobbyCode })}><Label>Start Game</Label></Button>
+    {#if auth.currentUser?.uid === lobby.uids[0]}<div class="actions">
+        <Button on:click={() => start()}><Label>Start Game</Label></Button>
+      </div>
+    {/if}
```
I don't like how the `<div>` is on the same line as `{#if}` but the `</div>` is on a diffrent line from `{/if}`

mrjoshua520 committed:
refactor(functions): 🔥 Remove unneeded awaits

nstringham committed:
Merge branch 'development' of https://github.com/cat-acronym-team/cat-me-if-you-can into jsalkeld/leave-lobby

nstringham reviewed (approved):
approving again after resolving merge conflicts
